Prevent orphaned image uploads via REST API when product creation fails due to validation errors

sunyatasattva opened the pull request:
### Submission Review Guidelines:

-   [x] I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   [x] I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   [x] I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR addresses an issue where images sent during a product creation request via the REST API (`/wc/v3/products`) were being downloaded and saved to the Media Library even if the product creation ultimately failed due to a validation error. This would happen (as exemplified in #58210) when for example a product with the same SKU existed but was in the trash. Each failed attempt would result in a new, unreferenced image attachment.

The fix now tracks any attachment IDs that are processed by `set_product_images()` during the request. If the request returns a `WP_Error` or an exception occurs, the newly tracked attachment IDs are looped through, and `wp_delete_attachment( $id, true )` is called to remove the orphaned images.

<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
Closes #58210

### How to test the changes in this Pull Request:

1.  Using a REST API client, make a `POST` request to `/wp-json/wc/v3/products` including this example body:

```
{
  "name": "Test Product",
  "type": "simple",
  "regular_price": "19.99",
  "sku": "duplicate-sku-test",
  "images": [
    {
      "src": "https://fastly.picsum.photos/id/107/536/354.jpg?hmac=fLKWiXSw_CxcT7QPAtGkCxVjxdQwZ4Xnl0a3d_ib9PA"
    }
  ]
}
```

2. Verify the product is created and the image is present in the WordPress Media Library.
3.  Trash this new product but don't permanently delete it.
4.  Make another `POST` request to `/wp-json/wc/v3/products` with the exact same body as above.
5. The API request should fail.
6. Go to the WordPress Media Library. Make sure no new image has been added.

### Testing that has already taken place:

-   A new PHPUnit test, `test_create_product_with_duplicate_sku_trashed_original_cleans_up_images`, has been added to `plugins/woocommerce/tests/php/includes/rest-api/Controllers/Version3/class-wc-rest-products-controller-tests.php`.
-   This test specifically sets up a product with an image, trashes it, then attempts to create a new product with the same SKU and a new image.
-   It asserts that the second product creation fails and, crucially, that the image associated with the failed attempt is cleaned up from the media library, while the original product's image remains.
-   Local manual testing following the steps above confirmed the fix.

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [x] Patch

#### Type

-   [x] Fix - Fixes an existing bug

#### Message
Prevent orphaned image uploads via REST API when product creation fails due to validation errors (e.g., duplicate SKU with original in trash).

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment
</details>

sunyatasattva committed:
Add cleanup for orphaned images on product creation failure

Enhance the `WC_REST_Products_Controller` to track processed
attachment IDs during product creation. If the creation fails,
the controller will now delete any orphaned images that were
uploaded during the request.

Fixes #58210

sunyatasattva committed:
Add unit tests

sunyatasattva committed:
Lint PHP

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

sunyatasattva committed:
Merge branch 'trunk' into fix/58210-clean-up-orphaned-images

dinhtungdu commented on plugins/woocommerce/includes/rest-api/Controllers/Version3/class-wc-rest-products-controller.php:
```diff
@@ -66,6 +66,13 @@ class WC_REST_Products_Controller extends WC_REST_Products_V2_Controller {
 	 */
 	private $exclude_status = array();
 
+	/**
+	 * Stores attachment IDs processed during the current request for potential cleanup.
+	 *
+	 * @var array
+	 */
+	private $processed_attachment_ids_for_request = array();
```
While this new property does the job, I wonder if we have other alternative to store the attachment IDs, for example [a request param](https://developer.wordpress.org/reference/classes/wp_rest_request/set_param/). The reasons are:
- Minimizing instance state.
- I think the processed attachment IDs naturally belong to the request data than the REST controller instance. The data is also belong to the request only, so you don't need to populate and cleanup between usage.

sunyatasattva replied:
I totally agree, I am not happy with this solution either.

But `set_product_images` is called from deep within the `prepare_object_for_database` method. The actual success or failure of the product creation is determined later.

Given the existing call chain (`create_item` → `prepare_object_for_database` → `set_product_images`), the private property is the least disruptive way I found to “tunnel” the state from the point of creation to the point where the rollback decision is made.

Alternatives I thought of (which I can get into, if you wish) would involve altering the fundamental signatures of parent methods or re-architecting the product creation flow. I'm not sure how you suggest using request params, but I'm all ears. I'm really happy if we can get rid of this.

dinhtungdu replied:
Oh, I thought `$request` was available for use through the class, but it's not. My bad again. 

> the private property is the least disruptive way I found to “tunnel” the state from the point of creation to the point where the rollback decision is made.

Yeah, I agree.

dinhtungdu commented on plugins/woocommerce/includes/rest-api/Controllers/Version3/class-wc-rest-products-controller.php:
```diff
@@ -574,13 +583,19 @@ protected function set_product_images( $product, $images ) {
 					}
 
 					$attachment_id = wc_rest_set_uploaded_image_as_attachment( $upload, $product->get_id() );
+					$is_new_upload = true;
```
Why don't we set the processed attachment ID here directly? At this point, the `$attachment_id` is always greater than `0`.

sunyatasattva replied:
Mm, I don't think that's the case, though. The point here is that `$attachment_id` can in fact be `0` at this point, in which case it triggers the upload logic. That's when the attachment gets assigned an id. In a sense that's the core of the fix. 🤔

dinhtungdu replied:
Right, my bad, `wp_insert_attachment` can still return `0`. False alarm! 🙈

dinhtungdu commented on plugins/woocommerce/tests/php/includes/rest-api/Controllers/Version3/class-wc-rest-products-controller-tests.php:
```diff
@@ -1496,6 +1496,85 @@ public function test_set_cogs_value_for_simple_product_via_post_request() {
+		);
+		$response_original_product = $this->server->dispatch( $request_original_product );
+
+		$this->assertEquals( 201, $response_original_product->get_status(), 'Failed to create the initial product with an image.' );
```
The product should be created successfully in the first step, right?

sunyatasattva replied:
Sorry, I'm not entirely familiar with PHPUnit, but I thought that the [`$message` argument is for when the assertion fails](https://docs.phpunit.de/en/10.5/assertions.html#assertequals), right?

dinhtungdu replied:
You're right, that message is for when the assertion fails, it's me that not familiar with PHPUnit test. It's not my best day to review a PR 😂.

dinhtungdu replied:
I want to comment on the custom messages in this test. The default error message uses the format of: Expecting something, got something else. It is very clear to me what the expectation is and what the actual result is. Maybe it's just me misreading those messages the first time, I didn't get them right before but now I do.

dinhtungdu reviewed (approved):
I manually tested creating duplicated trashed request on my end end the image is deleted after failed product creation. There is a minor comment about the test message. Pre-approving as this is LGTM already. Failed unit test doesn't look related.


#### 🖼️ Screenshots/ screen recordings
<img width="1800" alt="image" src="https://github.com/user-attachments/assets/030af346-5465-4604-b866-64404b42f0be" />

#### 🌐 Site details
<details><summary>System Status Report</summary>

```
### WordPress Environment ###

WordPress address (URL): [Redacted]
Site address (URL): [Redacted]
WC Version: 10.0.0-dev
Legacy REST API Package Version: The Legacy REST API plugin is not installed on this site.
Action Scheduler Version: ✔ 3.9.2
Log Directory Writable: ✔
WP Version: 6.8.1
WP Multisite: –
WP Memory Limit: 2 GB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: nginx/1.25.4
Server Architecture: Darwin 24.5.0 arm64
PHP Version: 7.4.33
PHP Post Max Size: 8 MB
PHP Time Limit: 30
PHP Max Input Vars: 1000
cURL Version: 8.5.0
OpenSSL/3.1.4

SUHOSIN Installed: –
MySQL Version: 8.0.33
Max Upload Size: 2 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

[REDACTED]

### Post Type Counts ###

attachment: 25
global_product_addon: 1
page: 17
post: 7
product: 19
product_tab: 1
product_variation: 7
revision: 155
shop_order_placehold: 1
wp_font_face: 36
wp_font_family: 12
wp_global_styles: 3
wp_navigation: 1
wp_template: 6
wp_template_part: 8

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ❌Error messages should not be shown to visitors.

### Active Plugins (4) ###

Laravel DD for Wordpress: by Peter Hegman – 1.0.1
WooCommerce Beta Tester: by WooCommerce – 3.0.0
WooCommerce: by Automattic – 10.0.0-dev
WP BASIC Auth: by wokamoto – 1.1.4

### Inactive Plugins (13) ###

Akismet Anti-spam: Spam Protection: by Automattic - Anti-spam Team – 5.3.6
bbPress: by The bbPress Contributors – 2.6.13
Block Hooks Experiments for WooCommerce: by Automattic –
Custom Product Tabs Manager: by Addify – 1.2.2
Gutenberg: by Gutenberg Team – 20.3.0
Hello Dolly: by Matt Mullenweg – 1.7.2
Product Filters for WooCommerce: by WooCommerce – 1.4.38
Query Monitor: by John Blackbourn – 3.17.2
What Can Wp Scripts Do: by The WordPress Contributors – 0.1.0
WooCommerce: by Automattic – 9.7.1 (update to version 9.8.5 is available)
WooCommerce PayPal Payments: by PayPal – 3.0.0 (update to version 3.0.6 is available)
WooCommerce Product Add-Ons: by Woo – 7.8.1 (update to version 7.9.0 is available)
WooPayments: by WooCommerce – 9.1.0 (update to version 9.4.0 is available)

### Must Use Plugins (1) ###

WooCommerce Product Details Extension Demo: by  – 1.0.0

### Settings ###

Legacy API Enabled: –
Force SSL: –
Currency: USD ($)
Currency Position: right_space
Thousand Separator: .
Decimal Separator: ,
Number of Decimals: 1
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –
Enforce Approved Product Download Directories: ✔
HPOS feature enabled: ✔
Order datastore: Automattic\WooCommerce\Internal\DataStores\Orders\OrdersTableDataStore
HPOS data sync enabled: –
Enabled Features: analytics
marketplace
order_attribution
site_visibility_badge
remote_logging
email_improvements
point_of_sale
custom_order_tables


### Logging ###

Enabled: ✔
Handler: Automattic\WooCommerce\Internal\Admin\Logging\LogHandlerFileV2
Retention period: 30 days
Level threshold: –
Log directory size: 409 KB

### WC Pages ###

Shop base: #7 - /shop/
Cart: #8 - /cart/ -  Contains the woocommerce/cart block
Checkout: #9 - /checkout/ -  Contains the woocommerce/checkout block
My account: #10 - /my-account/ -  Contains the [woocommerce_my_account] shortcode
Terms and conditions: ❌ Page not set

### Theme ###

Name: Twenty Twenty-Four
Version: 1.3
Author URL: https://wordpress.org
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
Theme type: Block theme
WooCommerce Support: ❌ Not declared

### Templates ###

Overrides: /Users/tung/workspace/woocommerce/plugins/woocommerce/templates/block-notices/error.php
/Users/tung/workspace/woocommerce/plugins/woocommerce/templates/block-notices/notice.php
/Users/tung/workspace/woocommerce/plugins/woocommerce/templates/block-notices/success.php


### Admin ###

Enabled Features: activity-panels
analytics
product-block-editor
experimental-blocks
coupons
core-profiler
customize-store
customer-effort-score-tracks
import-products-task
experimental-fashion-sample-products
shipping-smart-defaults
shipping-setting-tour
homescreen
marketing
minified-js
mobile-app-banner
onboarding
onboarding-tasks
pattern-toolkit-full-composability
payment-gateway-suggestions
product-custom-fields
printful
remote-inbox-notifications
remote-free-extensions
shipping-label-banner
subscriptions
store-alerts
transient-notices
woo-mobile-welcome
wc-pay-promotion
wc-pay-welcome-page
async-product-editor-category-field
launch-your-store
add-to-cart-with-options-stepper-layout
blockified-add-to-cart
disable-core-profiler-fallback

Disabled Features: product-data-views
experimental-iapi-mini-cart
experimental-iapi-runtime
coming-soon-newsletter-template
product-pre-publish-modal
settings
product-editor-template-system
use-wp-horizon
beta-tester-slotfill-examples

Daily Cron: ✔ Next scheduled: 2025-05-31 10:10:56 +00:00
Options: ✔
Notes: 74
Onboarding: completed

### Action Scheduler ###

Complete: 117
Oldest: 2025-04-29 15:50:21 +0000
Newest: 2025-05-30 10:53:20 +0000

Failed: 7
Oldest: 2025-03-05 03:00:25 +0000
Newest: 2025-05-20 11:45:14 +0000

Pending: 1
Oldest: 2025-05-31 10:27:35 +0000
Newest: 2025-05-31 10:27:35 +0000


### Status report information ###

Generated at: 2025-05-30 10:59:50 +00:00
```
</details>

sunyatasattva committed:
Merge branch 'trunk' into fix/58210-clean-up-orphaned-images

sunyatasattva committed:
Fix `$original_product_id` being undefined

sunyatasattva committed:
Remove unit tests

Unit tests were too stubborn and kept failing as the test
environment works differently and, for some unknown
reason, doesn't throw an instance of `WP_Error`.

After many hours of debugging, the only solution I found
included changing the controller code to do manual SQL
queries. Doing such a big change for the sake of making
the unit test for this edge case work was not worth.

So, though it's a shame, I reverted and removed the unit
tests completely.

sunyatasattva committed:
Merge branch 'trunk' into fix/58210-clean-up-orphaned-images

dinhtungdu commented:
@sunyatasattva I looked deeper at the test and found that the duplicated product was actually created. That's because `WC()->is_rest_api_request()` return `false`. Changing the REQUEST URI to `/wp-json/wc/v3/products` fixes the test for me.

```php
	/**
	 * @testdox The product POST endpoint properly cleans up orphaned images when product creation fails.
	 */
	public function test_create_product_with_duplicate_sku_trashed_original_cleans_up_images() {
		$_SERVER['REQUEST_URI'] = '/wp-json/wc/v3/products';
		$original_product_sku = 'DUPLICATE_SKU_TEST_TRASHED';
		// This image `src` is used in other product API tests, using here for consistency.
		$shared_image_src = 'https://tung.mystagingwebsite.com/wp-content/uploads/2025/03/beanie-2-768x768.jpg'; // I can't use the Cloudup image, don't know why. I also think we should use a local one instead of remote.

		// 1. Create the original product with its image.
		$request_original_product = new WP_REST_Request( 'POST', '/wc/v3/products' );
		$request_original_product->set_body_params(
			array(
				'name'          => 'Original Trashed Product',
				'sku'           => $original_product_sku,
				'type'          => 'simple',
				'regular_price' => '10',
				'images'        => array(
					array(
						'src' => $shared_image_src,
					),
				),
			)
		);
		$response_original_product = $this->server->dispatch( $request_original_product );

		$this->assertEquals( 201, $response_original_product->get_status(), 'Failed to create the initial product with an image.' );

		$original_product_data = $response_original_product->get_data();
		$original_product_id   = $original_product_data['id'];
		$original_attachment_id = $original_product_data['images'][0]['id'];

		// 2. Move the original product to trash.
		wp_trash_post( $original_product_id );

		$attachments_before_failed_attempt = count(
			get_posts(
				array(
					'post_type'   => 'attachment',
					'post_status' => 'inherit',
					'numberposts' => -1,
				)
			)
		);

		// 3. Attempt to create a new product with the same SKU and another image.
		$create_request_for_failure = new WP_REST_Request( 'POST', '/wc/v3/products' );
		$create_request_for_failure->set_body_params(
			array(
				'name'          => 'New Product Attempt That Fails',
				'sku'           => $original_product_sku, // Duplicate SKU.
				'type'          => 'simple',
				'regular_price' => '20',
				'images'        => array(
					array(
						'src' => $shared_image_src,
						'alt' => 'New Image To Be Cleaned Up',
					),
				),
			)
		);
		$failed_creation_response = $this->server->dispatch( $create_request_for_failure );
		$failed_creation_response_data = $failed_creation_response->get_data();

		$this->assertEquals( 400, $failed_creation_response->get_status(), 'Product creation attempt with duplicate SKU should return HTTP 400.' );
		$this->assertEquals( 'woocommerce_rest_product_not_created', $failed_creation_response_data['code'] );

		$attachments_after_failed_attempt = count(
			get_posts(
				array(
					'post_type'   => 'attachment',
					'post_status' => 'inherit',
					'numberposts' => -1,
				)
			)
		);

		$this->assertEquals( $attachments_before_failed_attempt, $attachments_after_failed_attempt, 'Number of attachments should remain unchanged after the failed product creation attempt, indicating cleanup of the second image.' );
		$this->assertNotNull( get_post( $original_attachment_id ), 'Original attachment for the initially created product should still exist.' );

		wp_delete_post( $original_product_id, true );
		wp_delete_attachment( $original_attachment_id, true );
	}
```

<img width="1025" alt="image" src="https://github.com/user-attachments/assets/01044e2d-8caf-498b-9360-a38d1dceef0b" />

sunyatasattva committed:
Revert "Remove unit tests"

This reverts commit 917db465a3345e16ba05d9a0d5dbe4382ace3eac.

sunyatasattva committed:
Manually override the `$_SERVER['REQUEST_URI']` in the unit test.

Kudos to @dihntungdu for discovering this. Closes WOOPLUG-4376.

sunyatasattva committed:
Linting

sunyatasattva committed:
Merge branch 'trunk' into fix/58210-clean-up-orphaned-images
